src/libsyntax/parse/parser.rs

```diff
@@ -233,6 +233,21 @@ macro_rules! maybe_whole {
     )
 }
 
+/// Uses $parse_expr to parse an expression and returns the span of the interpolated
+/// token or the span of the parsed expression, if it was not interpolated
+macro_rules! interpolated_or_expr_span {
```

Comment:
Your argument still doesn't make sense to me.

Both `Parser::parse_bottom_expr` and  `Parser::parse_prefix_expr` have the same return type, so its not like you are using the macro to inject some kind of duck-typing...

To be clear: You have a macro here that takes in two expressions as input (`$p` and `$parse_expr`). It unconditionally evaluates both expressions at the outset. (There's a second evaluation of `$p` but I don't think that was a deliberate choice.)

So, why wouldn't this work:

```rust
impl<'a> Parser<'a> {
    /// Uses $parse_expr to parse an expression and returns the span of the interpolated
    /// token or the span of the parsed expression, if it was not interpolated

    fn interpolated_or_expr_span(&self, expr: PResult<'a, P<Expr>>) -> PResult<'a, (Span, P<Expr>)> {
        let is_interpolated = self.token.is_interpolated();
        expr.map(|e| {
            if is_interpolated {
                (self.last_span, e)
            } else {
                (e.span, e)
            }
        })
    }
}
```

(And then update the macro calls to call this method instead.)